homeassistant/components/ddwrt/device_tracker.py

```diff
@@ -65,54 +77,93 @@ def scan_devices(self):
-            self.mac2name = {}
-            for idx in range(0, num_clients):
-                # The data is a single array
-                # every 5 elements represents one host, the MAC
-                # is the third element and the name is the first.
-                mac_index = (idx * 5) + 2
-                if mac_index < len(elements):
-                    mac = elements[mac_index]
-                    self.mac2name[mac] = elements[idx * 5]
+            # This update the mac2name object with the name of the specified
+            # device.
+            self.scan_table(device)
 
         return self.mac2name.get(device)
 
+    def scan_table(self, device, table='dhcp_leases'):
+        """Scan the router's ARP table to find the device's name."""
+        # If not initialised and not already scanned and not found.
+        url = '{}://{}/Status_Lan.live.asp'.format(
+            self.protocol, self.host)
+        data = self.get_ddwrt_data(url)
+
+        if not data:
+            return None
+
+        dhcp_leases = data.get(table, None)
+
+        if not dhcp_leases:
+            return None
+
+        # The data is a single array
+        # For 'dhcp_leases', every 5 elements represents one host,
+        # the MAC is the third element and the name is the first.
+        if table == 'dhcp_leases':
+            set_columns = 5
+
+        # For 'arp_table', every 4 elements represents one host,
+        # The position of the MAC and the name remains the same.
+        else:
+            set_columns = 4
+
+        # Remove leading and trailing quotes and spaces
+        cleaned_str = dhcp_leases.replace(
+            "\"", "").replace("\'", "").replace(" ", "")
+        elements = cleaned_str.split(',')
+        num_clients = int(len(elements) / set_columns)
+
+        self.mac2name = {}
+
+        for idx in range(0, num_clients):
+            mac_index = (idx * set_columns) + 2
+            # Make sure the index is not out of scope.
+            if mac_index < len(elements):
+                mac = elements[mac_index]
+
+                # Make sure the index is not out of scope.
+                if idx * set_columns < len(elements):
+                    name = elements[idx * set_columns]
+                    self.mac2name[mac] = name
+
+                    # If the name is a *, it means the name wasn't found in
+                    # the dhcp_lease table. Check on the arp_table instead.
+                    if name == '*' and table != 'arp_table':
+                        self.scan_table(device, 'arp_table')
+
+        return
+
     def _update_info(self):
         """Ensure the information from the DD-WRT router is up to date.
 
         Return boolean if scanning successful.
         """
-        _LOGGER.info("Checking ARP")
+        if self.type.lower() == 'wireless':
+            _LOGGER.info("Checking Wireless ARP")
+            url = 'http://{}/Status_Wireless.live.asp'.format(self.host)
```

Comment:
I think this PR has too many changes and additions to device protocol specific code which shouldn't be in home assistant.

We want the device specific interface code, I/O and data parsing, to be extracted to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#4-communication-with-devices-services

@balloob?

Reply:
I agree.